### ISSUE TITLE
Inconsistent Gradients with DDP and AMP when GradScaler Scales Differ Across Processes

### ISSUE THREAD

**Issue by hywang66** [pytorch/pytorch#136009]
### 🐛 Describe the bug

# Issue Description
I’m encountering an issue when using PyTorch’s DistributedDataParallel (DDP) in combination with Automatic Mixed Precision (AMP) and GradScaler. Specifically, when the GradScaler has different scaling factors across processes, the unscaled gradients become inconsistent after synchronization, leading to diverging model parameters across different ranks.

- **What happens:** After a few training steps, the parameters of the model diverge across different ranks.
- **Why it happens:** Each process uses a GradScaler with a different scaling factor. During the backward pass, DDP synchronizes the scaled gradients via all_reduce. After synchronization, each process unscales the gradients using its own scaling factor. Since the scaling factors differ, the unscaled gradients become inconsistent across processes.
- **Result:** The optimizer updates parameters differently on each process, leading to divergence in model parameters, which violates DDP’s assumption of synchronized parameters.

**Expected Behavior:**

- Model parameters should remain synchronized across all processes when using DDP, even when AMP and GradScaler are employed.
- Unscaled gradients should be consistent across processes after synchronization, ensuring consistent parameter updates.

# Reproduction Steps

Here is a minimal script that reproduces the issue. It should be run on a machine with at least 2 GPUs.

## Code
```
import torch
import torch.nn as nn
import torch.optim as optim
import torch.distributed as dist
import torch.multiprocessing as mp

def train(rank, world_size):
    # Initialize the process group
    dist.init_process_group(
        backend='nccl',
        init_method='tcp://127.0.0.1:12355',
        world_size=world_size,
        rank=rank
    )

    # Set the device for this process
    torch.cuda.set_device(rank)
    device = torch.device(f'cuda:{rank}')

    # Create the model and move it to the appropriate device
    model = nn.Linear(10, 1).to(device)

    # Wrap the model in DDP
    model = nn.parallel.DistributedDataParallel(model, device_ids=[rank])

    # Create optimizer
    optimizer = optim.SGD(model.parameters(), lr=0.1)

    # Create GradScaler with different init_scale for each rank
    if rank == 0:
        scaler = torch.cuda.amp.GradScaler(init_scale=65536.0)
    else:
        scaler = torch.cuda.amp.GradScaler(init_scale=131072.0)

    # Dummy input and target
    input = torch.randn(20, 10, device=device)
    target = torch.randn(20, 1, device=device)

    # Loss function
    criterion = nn.MSELoss()

    for step in range(20):
        optimizer.zero_grad()

        with torch.cuda.amp.autocast():
            output = model(input)
            loss = criterion(output, target)

        scaler.scale(loss).backward()
        scaler.step(optimizer)
        scaler.update()

        # Now, check if the model parameters are the same across ranks
        # Gather parameters to rank 0
        with torch.no_grad():
            for name, param in model.named_parameters():
                # Use CUDA tensors for NCCL backend
                param_cuda = param.detach().clone()
                param_list = [torch.zeros_like(param_cuda) for _ in range(world_size)]
                dist.all_gather(param_list, param_cuda)

                if rank == 0:
                    # Move parameters to CPU for comparison
                    param_list_cpu = [p.cpu() for p in param_list]
                    # Compare the parameters
                    params_equal = True
                    for i in range(1, world_size):
                        if not torch.allclose(param_list_cpu[0], param_list_cpu[i], atol=1e-6):
                            params_equal = False
                            print(f"Step {step}, Parameter '{name}': Divergence detected between rank 0 and rank {i}")
                            print(f"Difference: {torch.norm(param_list_cpu[0] - param_list_cpu[i])}")
                    if params_equal:
                        print(f"Step {step}, Parameter '{name}': Parameters are the same across all ranks.")

    # Clean up
    dist.destroy_process_group()

def main():
    # Get the world_size (number of GPUs)
    world_size = 2
    mp.spawn(train, args=(world_size,), nprocs=world_size, join=True)

if __name__ == "__main__":
    main()
```

## Results observed

> 
> Step 0, Parameter 'module.weight': Parameters are the same across all ranks.
> Step 0, Parameter 'module.bias': Parameters are the same across all ranks.
> Step 1, Parameter 'module.weight': Parameters are the same across all ranks.
> Step 1, Parameter 'module.bias': Parameters are the same across all ranks.
> Step 2, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.14353081583976746
> Step 2, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.02787170559167862
> Step 3, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.24856851994991302
> Step 3, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.050987616181373596
> Step 4, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.3269207179546356
> Step 4, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.07000923156738281
> Step 5, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.386452853679657
> Step 5, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.08551520109176636
> Step 6, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.4324363172054291
> Step 6, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.09806707501411438
> Step 7, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.4685640335083008
> Step 7, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.1081928014755249
> Step 8, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.49736350774765015
> Step 8, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.1163562536239624
> Step 9, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.5205991864204407
> Step 9, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.12285497784614563
> Step 10, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.5395646095275879
> Step 10, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.12800902128219604
> Step 11, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.5552190542221069
> Step 11, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.13209722936153412
> Step 12, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.5682567358016968
> Step 12, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.13533057272434235
> Step 13, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.5792085528373718
> Step 13, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.13783301413059235
> Step 14, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.5884769558906555
> Step 14, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.13978765904903412
> Step 15, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.5963413119316101
> Step 15, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.14129216969013214
> Step 16, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.6030936241149902
> Step 16, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.14240606129169464
> Step 17, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.608910858631134
> Step 17, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.14322392642498016
> Step 18, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.6139338612556458
> Step 18, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.1438220739364624
> Step 19, Parameter 'module.weight': Divergence detected between rank 0 and rank 1
> Difference: 0.6182945370674133
> Step 19, Parameter 'module.bias': Divergence detected between rank 0 and rank 1
> Difference: 0.14422795176506042
> 

### Versions

PyTorch version: 2.2.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux release 8.10 (Ootpa) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-22)
Clang version: Could not collect
CMake version: version 3.26.5
Libc version: glibc-2.28

Python version: 3.9.7 | packaged by conda-forge | (default, Sep 29 2021, 19:23:11)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-4.18.0-553.16.1.el8_10.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA RTX A5000
GPU 1: NVIDIA RTX A5000

Nvidia driver version: 550.90.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              48
On-line CPU(s) list: 0-47
Thread(s) per core:  1
Core(s) per socket:  24
Socket(s):           2
NUMA node(s):        2
Vendor ID:           AuthenticAMD
CPU family:          23
Model:               49
Model name:          AMD EPYC 7352 24-Core Processor
Stepping:            0
CPU MHz:             3097.491
CPU max MHz:         2300.0000
CPU min MHz:         1500.0000
BogoMIPS:            4600.08
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            16384K
NUMA node0 CPU(s):   0-23
NUMA node1 CPU(s):   24-47
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] onnxruntime==1.14.1
[pip3] pytorch-lightning==2.0.1
[pip3] pytorch-msssim==1.0.0
[pip3] torch==2.2.1
[pip3] torchaudio==2.0.0
[pip3] torchmetrics==0.11.4
[pip3] torchvision==0.17.1
[pip3] triton==2.2.0
[conda] blas                      1.0                         mkl
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py39h7f8727e_0
[conda] mkl_fft                   1.3.1            py39hd3c417c_0
[conda] mkl_random                1.2.2            py39h51133e4_0
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-cuda              11.7                 h778d358_3    pytorch
[conda] pytorch-lightning         2.0.1                    pypi_0    pypi
[conda] pytorch-msssim            1.0.0                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     2.2.1                    pypi_0    pypi
[conda] torchaudio                2.0.0                py39_cu117    pytorch
[conda] torchmetrics              0.11.4                   pypi_0    pypi
[conda] torchvision               0.17.1                   pypi_0    pypi
[conda] triton                    2.2.0                    pypi_0    pypi

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mcarilli @ptrblck @leslie-fang-intel @jgong5

**Comment by fegin** [comment]
DDP's reduction occurs when the gradient is ready, before the backward() function is finished. Therefore, DDP can only guarantee that the gradients are the same immediately after the backward() function is called. Beyond this point, DDP does not have control over how the gradients are modified. If you need to scale the gradients with different scale factors, you will need to synchronize the gradients as they have already passed the point where DDP can intercept the calls.

**Comment by hywang66** [comment]
Thank you for your reply! On the other hand, when the same initial scale factor is set, the gradient scalers on different GPUs should consistently maintain the same scale factor, ensuring the consistency of model gradients and weights. Is my understanding correct?

**Comment by fegin** [comment]
Yes, if the scale factor is the same, the results should be the same.
